### DISCUSSION TITLE
MVTLayer with static PBF/MVT tiles

### DISCUSSION THREAD

**Discussion by cornhundred** [visgl/deck.gl#5547] [Q&A] [answered]
I would like to try to load static PBF files (or MVT files) from GitHub into a deck.gl MVTLayer (if possible). I found this example where they load static PBF files from GitHub into a leaflet map - https://github.com/xycarto/static.vector.tiles and I want to know if it is possible to load something like this into the MVTLayer. 

I also got the MVT example from the documentation into ObservableHQ (https://observablehq.com/@cornhundred/deck-gl-mvt-layer), but I am unable to load the precomputed pbf files from the static.vector.tiles example into the MVTLayer. Is this type of thing possible and if so are there any examples of generating a custom vector tile map (e.g. using tippecanoe) and hosting on a service like GitHub? The idea of using the MVTLayer was suggested here https://github.com/visgl/deck.gl/discussions/5519 by @kylebarron.

**Comment by cornhundred** [comment, marked as answer]
I was able to get a custom built MVT dataset (built following the [tippiecanoe Rhode Island example](https://github.com/mapbox/tippecanoe#discontinuous-polygon-features-buildings-of-rhode-island-visible-at-all-zoom-levels)) working on [ObservableHQ](https://observablehq.com/@cornhundred/deck-gl-mvt-layer-custom) by using mapbox to host the MVT data as a public tileset. However, would it be possible to host the MVT tileset on GitHub or S3?

**Reply by Pessimistress** [reply]
Yes. Have you tried it? Are you seeing any errors?

The minimap example uses GitHub to host mvt tiles: https://github.com/visgl/deck.gl/blob/master/examples/website/radio/app.js#L24

**Reply by cornhundred** [reply]
Thanks @Pessimistress. Yes, I tried converting the single Rhode Island mvt file into pbf files (using `mb-util` as show [here](https://github.com/kylebarron/all-transit/tree/01e5eda5dcfffb200e9922db219eff5c173749a1#vector-tiles-for-operators-routes-stops) and hosted them on [GitHub here](https://github.com/cornhundred/vector_tile_pbf_test/tree/master/RhodesIsland)) and it did not work but I will have to try again to see which specific errors I was getting. 

I will also checkout the radio minimap example to see how it works.

**Reply by kylebarron** [reply]
I believe tippecanoe creates gzipped tiles by default, so you might have to unzip them to store them somewhere where you can't set the content header

**Reply by cornhundred** [reply]
@Pessimistress I made this fork of the ObservableHQ RhodeIsland notebook that tries to load the PBF files from GitHub instead of Mapbox 

https://observablehq.com/@cornhundred/deck-gl-mvt-layer-custom-github-pbf-files

I'm not seeing anything show up, but some of the PBF files appear to be loading correctly based on the 200 network responses 

<img width="367" alt="Screen Shot 2021-02-28 at 1 30 55 PM" src="https://user-images.githubusercontent.com/8352840/109429333-42bcf800-79c9-11eb-919a-f1a731c5917f.png">

Thanks @kylebarron, I'll look into this gzipping

**Reply by kylebarron** [reply]
Well the 404s correspond to files that don't exist https://github.com/cornhundred/vector_tile_pbf_test/tree/master/RhodesIsland/8/76

Do you get console errors?

**Reply by cornhundred** [reply]
@kylebarron, yes I saw some console errors shown below 

<img width="757" alt="Screen Shot 2021-02-28 at 1 43 04 PM" src="https://user-images.githubusercontent.com/8352840/109429671-f2469a00-79ca-11eb-8736-84a3ca88607c.png">

but the working example, hosted on mapbox, also has similar errors 

<img width="751" alt="Screen Shot 2021-02-28 at 1 43 54 PM" src="https://user-images.githubusercontent.com/8352840/109429691-0f7b6880-79cb-11eb-8e8a-a3166c1bf936.png">

**Reply by kylebarron** [reply]
Yes, you get 404 errors where tiles don't exist. That's expected. You also have separate `Unimplemented type: 3` errors, from the data that is loaded, I believe that's because the browser doesn't know to ungzip the data.

**Reply by cornhundred** [reply]
Ok I can try to unzip them. Do you know if there a way for tippecanoe to unzip them or a simple way to convert them all, or a way to tell deck to unzip?

**Reply by kylebarron** [reply]
You can use `--no-tile-compression`. Or just use the command line to unzip https://github.com/mapbox/tippecanoe/pull/386, https://github.com/mapbox/tippecanoe/issues/582, https://github.com/mapbox/tippecanoe/issues/383.

**Reply by cornhundred** [reply]
Great, @kylebarron!! I'll try that soon and let you know how it works

**Reply by cornhundred** [reply]
@kylebarron adding the `--no-tile-compression` command to tippecanoe worked. I ran 

```
tippecanoe -zg -o RhodeIslandDC.mbtiles --drop-densest-as-needed --extend-zooms-if-still-dropping --no-tile-compression RhodeIsland.geojson
```

Then ran mb-util to generate the tile pyramid directory

```
mb-util RhodeIslandDC.mbtiles pbf_tiles/RhodesIslandDC --image_format=pbf
```

The files are being hosted on this public repo:

https://github.com/cornhundred/vector_tile_pbf_test/tree/master/RhodesIslandDC

and the example is on ObservableHQ: https://observablehq.com/@cornhundred/deck-gl-mvt-layer-custom

Next, I want to try to load private data from a GitHub repo by passing the custom header as done here for the TileLayer https://github.com/visgl/deck.gl/discussions/5499

Thanks again everyone for the help :D
